### PR TITLE
Feature 53 create tags

### PR DESCRIPTION
**Pull request by sterlinglaird**
Tried to make this pull request earlier but it got deleted for some reason.

This issue is still a work in progress and will need more testing and potentially some design reconsiderations, suggestions are very much welcome!

When in the lowest zoom level (third level), right click on a commit to bring up a newly added context menu which currently only has the option to create a new tag. In the future we can add other functionality to this drop down. It is a little unintuitive to need to be on the lowest zoom level, but that is because in higher zoom levels a node actually represents a collection (or may just be one) of commits while the third level actually represents one commit per node. In the 'Create Tag' menu you can choose the tag name, whether it is an annotated or lightweight tag (and provide an optional tag message), and whether or not to push the tag upstream. After a successful tag creation, the repo is reloaded and the graph is updated to reflect changes that are implemented in #54. Pushing tags multiple times crashes for the same reason as #71 so changes from that may need to be incorporated in this issue.

### CODE REVIEW

**Comment by SiRumCz** [comment]
image

Is it better to have the display message here be something like "successfully pushed or created tags xxxxx"?

Adding and pushing lightweight and annotated tag to remote or local repo worked for me, except the app crashed after I played with it couple times.

**Comment by sterlinglaird** [comment]
> image
> 
> Is it better to have the display message here be something like "successfully pushed or created tags xxxxx"?
> 
> Adding and pushing lightweight and annotated tag to remote or local repo worked for me, except the app crashed after I played with it couple times.

Good idea. What was the nature of the crashes? I haven't seen any (at least related to this..) but maybe theres other code paths I didn't think of. The app will crash when pushing multiple times though as identified in #71.

File: app/misc/git.ts [hunk 1]
```diff
@@ -616,6 +616,61 @@ function mergeLocalBranches(element) {
...
+          .then(function (remotes) {
+            return repo.getRemote(remotes[0]);
+          })
+          .then(function(remote){
+            return remote.push(
+              ["refs/tags/" + tagName + ":refs/tags/" + tagName],
+              {
+                callbacks: {
+                  credentials: function () {
+                    return cred;
```
Comment (sterlinglaird) [inline app/misc/git.ts]: This needs to be updated with the new function added with #71

**Comment by SiRumCz** [comment]
Then I suppose I did few pushes. You probably wanna inverstigate on that one.

**Comment by sterlinglaird** [comment]
> Then I suppose I did few pushes. You probably wanna investigate on that one.

An other team is investigating/solving it in #82 /#71. I will incorporate the fix once it merges to master.

**Pull request merged**